libs/base/Data/List1.idr

```diff
@@ -141,3 +143,65 @@ Ord a => Ord (List1 a) where
 export
 consInjective : the (List1 a) (x ::: xs) === (y ::: ys) -> (x === y, xs === ys)
 consInjective Refl = (Refl, Refl)
+
+------------------------------------------------------------------------
+-- Zip/Unzip
+
+export
+zipWith : (a -> b -> c) -> List1 a -> List1 b -> List1 c
+zipWith f (x ::: xs) (y ::: ys) = f x y ::: zipWith' xs ys
+  where
+    zipWith' : List a -> List b -> List c
+    zipWith' [] _ = []
+    zipWith' _ [] = []
+    zipWith' (x :: xs) (y :: ys) = f x y :: zipWith' xs ys
+
+export
+zip : List1 a -> List1 b -> List1 (a, b)
+zip = zipWith \x, y => (x, y)
+
+export
+zipWith3 : (a -> b -> c -> d) -> List1 a -> List1 b -> List1 c -> List1 d
+zipWith3 f (x ::: xs) (y ::: ys) (z ::: zs) = f x y z ::: zipWith3' xs ys zs
+  where
+    zipWith3' : List a -> List b -> List c -> List d
+    zipWith3' [] _ _ = []
+    zipWith3' _ [] _ = []
+    zipWith3' _ _ [] = []
+    zipWith3' (x :: xs) (y :: ys) (z :: zs) = f x y z :: zipWith3' xs ys zs
+
+export
+zip3 : List1 a -> List1 b -> List1 c -> List1 (a, b, c)
+zip3 = zipWith3 \x, y, z => (x, y, z)
```

Comment:
Note that you can use sections i.e. `(,,)`.

Reply:
Applied this suggestion in `Data.List`, `Data.List1` and `Data.Stream`